pkg/operator/hive/manageddns.go

```diff
@@ -37,38 +42,105 @@ func (r *ReconcileHiveConfig) teardownLegacyExternalDNS(hLog log.FieldLogger) er
+// return the current configmap of the current deployment if the settings it contains match the desired settings.
+// It will also return the old configmap if a new one was created (so that it can be garbage collected).
+func (r *ReconcileHiveConfig) configureManagedDomains(logger log.FieldLogger, instance *hivev1.HiveConfig) (*corev1.ConfigMap, error) {
+	var mdConfigMap *corev1.ConfigMap
 
 	domains, err := json.Marshal(instance.Spec.ManagedDomains)
 	if err != nil {
-		return errors.Wrap(err, "failed to marshal managed domains list into the configmap")
+		return nil, errors.Wrap(err, "failed to marshal managed domains list into the configmap")
 	}
 
-	cm.Data = map[string]string{"managed-domains": string(domains)}
-	_, err = h.ApplyRuntimeObject(cm, scheme.Scheme)
-	return errors.Wrap(err, "error applying managed domains configmap")
+	newConfigMapData := map[string]string{managedDomainsConfigMapKey: string(domains)}
+
+	currentConfigMap, err := r.getCurrentConfigMap(newConfigMapData, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	if !reflect.DeepEqual(newConfigMapData, currentConfigMap.Data) {
+		log.Debug("need a new/updated configmap for managed domains")
+		mdConfigMap = &corev1.ConfigMap{}
+		mdConfigMap.Kind = "ConfigMap"
+		mdConfigMap.APIVersion = "v1"
+		mdConfigMap.GenerateName = managedDomainsConfigMapNamePrefix
+		mdConfigMap.Namespace = constants.HiveNamespace
+		mdConfigMap.Labels = map[string]string{configMapLabel: "true"}
+
+		mdConfigMap.Data = newConfigMapData
+
+		if err := r.Create(context.TODO(), mdConfigMap); err != nil {
+			return nil, errors.Wrap(err, "failed to save new managed domains configmap")
+		}
+		log.WithField("configmap", fmt.Sprintf("%s/%s", mdConfigMap.Namespace, mdConfigMap.Name)).
+			Debug("saved configmap for managed domains")
+
+	} else {
+		// If configmap data was equal keep using current configmap.
+		log.WithField("configmap", fmt.Sprintf("%s/%s", currentConfigMap.Namespace, currentConfigMap.Name)).
+			Info("using existing manage domains config map")
+		mdConfigMap = currentConfigMap
+	}
+
+	return mdConfigMap, nil
+}
+
+// getCurrentConfigMap will see if any existing configmap (for managed domains) already has the necessary
+// settings. It will also delete any configmaps (for managed domains) that have out-of-date contents
+// (so that the configmaps are not orphaned as config changes happen).
+func (r *ReconcileHiveConfig) getCurrentConfigMap(cmData map[string]string, logger log.FieldLogger) (*corev1.ConfigMap, error) {
+	configMapList := &corev1.ConfigMapList{}
+	labelSelector := map[string]string{configMapLabel: "true"}
+
+	currentConfigMap := &corev1.ConfigMap{}
+
+	err := r.List(context.TODO(), configMapList, client.MatchingLabels(labelSelector))
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to list config maps for managed domains")
+	}
+
+	for i, cm := range configMapList.Items {
```

Comment:
This leaves multiple configmaps if there are multiple that have the correct, current data. Coming out of this function, there should only be a single configmap.

Reply:
i didn't see a way that the operator could end up with multiple configmaps with correct, current data (unless someone comes in external to the code and starts making these configmaps) (or maybe the operator has multiple replicas running)

Reply:
I agree that it is unlikely. However, you are already looping through all of the configmaps so I figured it wouldn't be too hard to also ensure that we don't have duplicates. It is fine as is.

Reply:
I changed the flow to catch this case.